oneflow/core/job_rewriter/fuse_update_ops_pass.cpp

```diff
@@ -0,0 +1,167 @@
+/*
+Copyright 2020 The OneFlow Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+#include "oneflow/core/job_rewriter/op_graph_pass.h"
+#include "oneflow/core/register/runtime_blob_desc.h"
+#include "oneflow/core/framework/framework.h"
+
+namespace oneflow {
+
+namespace {
+
+std::function<bool(const OpNode* op_node)> MakePredicatorIsSafeToDelete(const OpGraph& op_graph) {
+  HashSet<std::string> ctrl_in_op_names;
+  op_graph.ForEachNode([&](const OpNode* op_node) {
+    for (const std::string& ctrl_in_op_name : op_node->op().op_conf().ctrl_in_op_name()) {
+      ctrl_in_op_names.insert(ctrl_in_op_name);
+    }
+  });
+  return [=](const OpNode* op_node) {
+    if (!op_node->op().op_conf().ctrl_in_op_name().empty()) { return false; }
+    if (ctrl_in_op_names.find(op_node->op().op_conf().name()) != ctrl_in_op_names.end()) {
+      return false;
+    }
+    return true;
+  };
+}
+
+bool IsUserOpWithTypeName(const OperatorConf& op_conf, const std::string& op_type_name) {
+  return op_conf.has_user_conf() && op_conf.user_conf().op_type_name() == op_type_name;
+};
+
+class FuseUpdateOpsPass final : public OpGraphPass {
+ public:
+  FuseUpdateOpsPass() = default;
+  ~FuseUpdateOpsPass() override = default;
+
+  bool IsEnabled() const override {
+    return GlobalJobDesc().job_conf().enable_fuse_model_update_ops();
+  }
+  Maybe<void> Apply(const OpGraph& op_graph, JobBuilder* job_builder) const override;
+};
+
+Maybe<void> FuseUpdateOpsPass::Apply(const OpGraph& op_graph, JobBuilder* job_builder) const {
+  const auto IsSafeToDelete = MakePredicatorIsSafeToDelete(op_graph);
+  op_graph.ForEachNode([&](const OpNode* op_node) {
+    if (!op_node->op().op_conf().has_user_conf()) { return; }
+    const user_op::UserOpConfWrapper user_op_conf(op_node->op().op_conf());
+    if (user_op_conf.op_type_name() != "sgd_update"
+        && user_op_conf.op_type_name() != "momentum_update"
+        && user_op_conf.op_type_name() != "adam_update") {
+      return;
+    }
+    if (user_op_conf.attr<float>("scale") != 1.0f || user_op_conf.attr<float>("l1") != 0.0f
+        || user_op_conf.attr<float>("l2") != 0.0f) {
+      return;
+    }
+    float l1 = 0;
+    float l2 = 0;
+    float scale = 1;
+    bool fused = false;
+    LogicalBlobId model_diff_lbi = GenLogicalBlobId(user_op_conf.input("model_diff", 0));
+
+    [&]() {
```

Comment:
// Do no change to  scope { } because `return` is used in the lambda for bypassing